splink/spark/spark_base.py

```diff
@@ -7,15 +7,37 @@ def size_array_intersect_sql(col_name_l, col_name_r):
     return f"size(array_intersect({col_name_l}, {col_name_r}))"
 
 
-def datediff_sql(col_name_l, col_name_r, date_threshold, date_metric):
-    if date_metric == "day":
-        date_f = f"abs(datediff({col_name_l}, {col_name_r}))"
-    elif date_metric in ["month", "year"]:
-        date_f = f"ceil(abs(months_between({col_name_l}, {col_name_r})"
-        if date_metric == "year":
-            date_f += " / 12))"
-        else:
-            date_f += "))"
+def datediff_sql(
+    col_name_l,
+    col_name_r,
+    date_threshold,
+    date_metric,
+    cast_str=False,
+    date_format=None,
+):
+    if date_format is None:
+        date_format = "yyyy-MM-dd"
+
+    if cast_str:
+        if date_metric == "day":
+            date_f = f"""abs(datediff(to_timestamp({col_name_l},
+            '{date_format}'),to_timestamp({col_name_r},'{date_format}')))"""
+        elif date_metric in ["month", "year"]:
+            date_f = f"""floor(abs(months_between(to_timestamp({col_name_l},
+            '{date_format}'),to_timestamp({col_name_r}, '{date_format}'))"""
+            if date_metric == "year":
+                date_f += " / 12))"
+            else:
+                date_f += "))"
+    else:
+        if date_metric == "day":
+            date_f = f"abs(datediff({col_name_l}, {col_name_r}))"
+        elif date_metric in ["month", "year"]:
+            date_f = f"ceil(abs(months_between({col_name_l}, {col_name_r})"
+            if date_metric == "year":
+                date_f += " / 12))"
+            else:
+                date_f += "))"
```

Comment:
Sorry, I've been nosey and had a quick poke around this PR.

I think the following may be slightly cleaner/easier to read:
```suggestion
    if cast_str:
        col_name_l = f"to_timestamp({col_name_l}, '{date_format}')"
        col_name_r = f"to_timestamp({col_name_r}, '{date_format}')"

    if date_metric == "day":
        date_f = f"abs(datediff({col_name_l}, {col_name_r}))"
    elif date_metric in ["month", "year"]:
        date_f = f"ceil(abs(months_between({col_name_l}, {col_name_r})"
        if date_metric == "year":
            date_f += " / 12))"
        else:
            date_f += "))"
```

Reply:
I agree- this is a lot nicer and thanks for the suggestion! How do I integrate this change/ is there a way to reopen the pull request?